jbinding-cpp/JBindingTools.cpp

```diff
@@ -33,6 +37,101 @@ static struct {
+        JBindingSession::_classLoaderObjects.insert(std::make_pair(JAVA_INTEGER, addClassLoaderObject(env, JAVA_INTEGER)));
+        JBindingSession::_classLoaderObjects.insert(std::make_pair(JAVA_LONG, addClassLoaderObject(env, JAVA_LONG)));
+        JBindingSession::_classLoaderObjects.insert(std::make_pair(JAVA_FLOAT, addClassLoaderObject(env, JAVA_FLOAT)));
+        JBindingSession::_classLoaderObjects.insert(std::make_pair(JAVA_DOUBLE, addClassLoaderObject(env, JAVA_DOUBLE)));
+        JBindingSession::_classLoaderObjects.insert(std::make_pair(JAVA_BOOLEAN, addClassLoaderObject(env, JAVA_BOOLEAN)));
+        JBindingSession::_classLoaderObjects.insert(std::make_pair(JAVA_STRING, addClassLoaderObject(env, JAVA_STRING)));
+        JBindingSession::_classLoaderObjects.insert(std::make_pair(JAVA_THROWABLE, addClassLoaderObject(env, JAVA_THROWABLE)));
+        JBindingSession::_classLoaderObjects.insert(std::make_pair(JAVA_DATE, addClassLoaderObject(env, JAVA_DATE)));
+
+#ifdef NATIVE_JUNIT_TEST_SUPPORT
+        JBindingSession::_classLoaderObjects.insert(std::make_pair("net/sf/sevenzipjbinding/junit/jbindingtools/Callback1", addClassLoaderObject(env, "net/sf/sevenzipjbinding/junit/jbindingtools/Callback1")));
+        JBindingSession::_classLoaderObjects.insert(std::make_pair("net/sf/sevenzipjbinding/junit/jbindingtools/JBindingTest", addClassLoaderObject(env, "net/sf/sevenzipjbinding/junit/jbindingtools/JBindingTest")));
+        JBindingSession::_classLoaderObjects.insert(std::make_pair("net/sf/sevenzipjbinding/junit/jbindingtools/ExceptionHandlingTest", addClassLoaderObject(env, "net/sf/sevenzipjbinding/junit/jbindingtools/ExceptionHandlingTest")));
+
+        JBindingSession::_classLoaderObjects.insert(std::make_pair("net/sf/sevenzipjbinding/junit/jnitools/JTestAbstractClass", addClassLoaderObject(env, "net/sf/sevenzipjbinding/junit/jnitools/JTestAbstractClass")));
+        JBindingSession::_classLoaderObjects.insert(std::make_pair("net/sf/sevenzipjbinding/junit/jnitools/JTestFinalClass", addClassLoaderObject(env, "net/sf/sevenzipjbinding/junit/jnitools/JTestFinalClass")));
+        JBindingSession::_classLoaderObjects.insert(std::make_pair("net/sf/sevenzipjbinding/junit/jnitools/Interface1", addClassLoaderObject(env, "net/sf/sevenzipjbinding/junit/jnitools/Interface1")));
+        JBindingSession::_classLoaderObjects.insert(std::make_pair("net/sf/sevenzipjbinding/junit/jnitools/ParamSpecTest", addClassLoaderObject(env, "net/sf/sevenzipjbinding/junit/jnitools/ParamSpecTest")));
+#endif
+    }
+    return JNI_VERSION_1_6;
+}
+
+jclass findClass(JNIEnv* env, const char* name) {
+    if (env->ExceptionCheck()) {
+        env->ExceptionClear();
+    }
+    jstring string = env->NewStringUTF(name);
+    jclass clazz = static_cast<jclass>(env->CallObjectMethod(JBindingSession::_classLoaderObjects.at(name), JBindingSession::_classLoaderID, string));
+    env->DeleteLocalRef(string);
+    return clazz;
+}
+#endif
+
```

Comment:
Workaround due to the Java ClassLoader not being accessible from native threads on Android.
Reference: [FindClass from any thread in Android JNI](https://exceptionshub.com/findclass-from-any-thread-in-android-jni.html)

Reply:
Very interesting... I just implemented a completely new classloader handling (not commited it yet) due to problems if using fatjar or spring bootstrap. The problems occur, if the system classloader can't load SevenZip class.

In order to overcome this, I stopped using FindClass altogether, using provided ClassLoader instead. This change may actually prove this pull request obsolete or may require some additional tweeks to it.

Unfortunately, my commit results in sporadic JVM crashes, that are very hard to debug. I will push it to a new branch for you to investigate.

On the other hand, maybe we can drop my commit and use this pull request, adopting it to the regular JVM (non-Android).

Reply:
AND! Thank you very much for your contribution of course! Sorry for not beginning with that!

Reply:
Here is my classloader branch: https://github.com/borisbrodski/sevenzipjbinding/tree/classloader-rebuild

Reply:
Tested the 'loadClass()' based solution from [classloader-rebuild](https://github.com/borisbrodski/sevenzipjbinding/commits/classloader-rebuild), but this does not appear to be working well on Android.

For example, when running `compression.CompressExceptionGetItemInformationTest`:
```
I/TRACE: env->FindClass() for 7-Zip class net/sf/sevenzipjbinding/IOutCreateCallback
I/TRACE: Canonical class name: 'net.sf.sevenzipjbinding.IOutCreateCallback'
I/TRACE: Clazz: '0x79'
I/TRACE: Exception: '0x0'
I/TRACE: env->FindClass() for 7-Zip class net/sf/sevenzipjbinding/IInStream
I/TRACE: Canonical class name: 'net.sf.sevenzipjbinding.IInStream'
A/ipjbinding.tes: java_vm_ext.cc:542] JNI DETECTED ERROR IN APPLICATION: JNI NewStringUTF called with pending exception java.lang.ClassNotFoundException: Didn't find class "net.sf.sevenzipjbinding.IInStream" on path: DexPathList[[directory "."],nativeLibraryDirectories=[/system/lib64, /vendor/lib64, /system/lib64, /vendor/lib64]]
A/ipjbinding.tes: java_vm_ext.cc:542]   at java.lang.Class dalvik.system.BaseDexClassLoader.findClass(java.lang.String) (BaseDexClassLoader.java:134)
A/ipjbinding.tes: java_vm_ext.cc:542]   at java.lang.Class java.lang.ClassLoader.loadClass(java.lang.String, boolean) (ClassLoader.java:379)
A/ipjbinding.tes: java_vm_ext.cc:542]   at java.lang.Class java.lang.ClassLoader.loadClass(java.lang.String) (ClassLoader.java:312)
A/ipjbinding.tes: java_vm_ext.cc:542] 
A/ipjbinding.tes: java_vm_ext.cc:542]     in call to NewStringUTF
...
```

Not sure why there's an exception pending for finding `net.sf.sevenzipjbinding.IInStream` when calling `NewStringUTF ()` though,, since this is _before_ calling `loadClass()`.

Android documentation does suggest it being possible to use `loadClass()` from non-Java based native threads, but this "requires some effort" (solution 3):
[FAQ: Why didn't FindClass find my class?](https://developer.android.com/training/articles/perf-jni#faq:-why-didnt-findclass-find-my-class)
This pull request implements solution 1 "Do your FindClass lookups once, in JNI_OnLoad".

Reply:
Thank you very much for the commit review! It's really great, if someone looks over your code finding such dumm mistakes :)

The fix the problem above may be just calling ExceptionClear() at the right place. So how do you run tests on Android?

Another question is: should we get classloader-rebuild branch to work with Android or discard it and make your pull request working for regular Java? What is your opinion on this?

Reply:
> The fix the problem above may be just calling ExceptionClear() at the right place.

True, any pending JNI exceptions should be cleared first. But the question is why the "java.lang.ClassNotFoundException" was generated for "net.sf.sevenzipjbinding.IInStream" to begin with. Also, if such an exception ever occurred, the library would not be working correctly anymore.

> So how do you run tests on Android?

I created a branch here, which includes the core of the `loadClass()` changes: [classloader-rebuild](https://github.com/omicronapps/7-Zip-JBinding-4Android/tree/classloader-rebuild). I only included the minimum amount of changes from the original classloader-rebuild branch required to get this to build and run on Android.
To run tests on Android, check this branch out locally, open the project in Android Studio, and run "AllTestSuite" or any individual unit test. For additional details, feel free to email me.

> Another question is: should we get classloader-rebuild branch to work with Android or discard it and make your pull request working for regular Java? What is your opinion on this?

Either way is fine, as long as a working solution can be found. At this point, I can not get classloader-rebuild to work on Android, as something fundamental appears broken or missing in that implementation. Possibly something specific is missing for Android, but if so I'm not sure what that may be.
As for the existing JNI_OnLoad solution (this pull request), this is already confirmed to be working reliably by several Android developers. I have not confirmed how this would run on a standard Java VM though.

Reply:
Ok, I think it's much easier to just discard my last commit by one and go with your pull requests. I will see later, how I can add this feature without breaking the android support.

Reply:
I push everything to the new "android" branch, so we can make it shine first :)

Reply:
Why are you checking  `if (pjvm->GetEnv((void**)&env, JNI_VERSION_1_6) == JNI_OK) {` ?
What if the version is greater than 1.6? It should work anyway, right?

Reply:
> Why are you checking if `(pjvm->GetEnv((void**)&env, JNI_VERSION_1_6) == JNI_OK) {` ?

Android currently supports JNI 1.6 across all OS versions. So Android applications will typically acquire a JNIEnv instance for JNI 1.6, like documented here: https://developer.android.com/training/articles/perf-jni#native-libraries

> What if the version is greater than 1.6? It should work anyway, right?

No, not supported, and will result in the application crashing. For example if using `JNI_VERSION_1_8` (0x00010008):
```
2021-01-01 12:00:00.000 12345-12345/? E/e.myapplicatio: Bad JNI version passed to GetEnv: 65544
2021-01-01 12:00:00.000 12345-12345/? D/AndroidRuntime: Shutting down VM

    --------- beginning of crash
2021-01-01 12:00:00.000 12345-12345/? E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.example.myapplication, PID: 12345
    java.lang.UnsatisfiedLinkError: JNI_ERR returned from JNI_OnLoad in "/data/app/com.example.myapplication-bl8Ny_peaK5e_v9hWSvnzw==/lib/arm64/libnative-lib.so"
```

Reply:
Ah, very good. Let's stick with JNI_VERSION_1_6 for now.

My idea is to remove all or almost all `#ifdef __ANDROID_API__`, so that we have one code base for Android and non-Android JVMs. After I/we get this to work, I would like to refactor all the code. It's not well structured and though hard to maintain. Also my clangd language server shows lot's of error in almost all h-files.

Do you think, it's overall a good idea to go this way?

In process of doing so I got some questions:

- Why are you storing in `JBindingSession::_classLoaderObjects` classloaders and not `jclass`es ?
- The type of `JBindingSession::_classLoaderObjects` is a map without comparator. This leads to problems. Is it on purpose or is it a bug?
(I added `std::str_cmp` for now, so that the key is the actual string and not the `char const *` pointer itself. In the future I would propose to use an enumeration for this.)
- I managed to run JUnit tests on my Android device (that are failing for now). Is it a possibility to run JUnit tests (on a Simulator or on a Device) in a headless mode? Should you then get also the console output of the tests?

Thank you very much for your support!

Reply:
Ok, I can't use your approach with classloader for some weird reason. I'm getting following exception:

```
"java.lang.LinkageError: loader (instance of  sun/misc/Launcher$AppClassLoader): attempted  duplicate class definition for nam
e: "net/sf/sevenzipjbinding/ICryptoGetTextPassword""
```

The meta code, that raises that exception is something like:

```
c = env->FindClass("net/sf/sevenzipjbinding/ICryptoGetTextPassword")
c.class.getClassLoader().findClass("net.sf.sevenzipjbinding.ICryptoGetTextPassword")
```

Any ideas, why it happen?


Also, in you original code

```
jclass findClass(JNIEnv* env, const char* name) {
    if (env->ExceptionCheck()) {
        env->ExceptionClear();
    }
    jstring string = env->NewStringUTF(name);
    jclass clazz = static_cast<jclass>(env->CallObjectMethod(JBindingSession::_classLoaderObjects.at(name), JBindingSession::_classLoaderID, string));
    env->DeleteLocalRef(string);
    return clazz;
}
```

you don't convert `name` to dot-notation: java/lang/Object => java.lang.Object. Why is it working on Android anyway?

Reply:
> My idea is to remove all or almost all #ifdef __ANDROID_API__, so that we have one code base for Android and non-Android JVMs. After I/we get this to work, I would like to refactor all the code. It's not well structured and though hard to maintain. Also my clangd language server shows lot's of error in almost all h-files.
> Do you think, it's overall a good idea to go this way?

Yes, removing all instances of `__ANDROID_API__`, where possible, would be a good first step. In particular all cases of deleting local JNI references should reasonably work across both Java and Android.

> Why are you storing in JBindingSession::_classLoaderObjects classloaders and not jclasses ?

`JNI_OnLoad()` ensures access to the ClassLoader, which is not available in native threads. So caching the ClassLoader object _for that particular class_ here will allow you to safely call that ClassLoader function from a native thread.

But I think this may very well be possible to further to simplify and optimize. However, the guidance provided in Android documentation ([FAQ: Why didn't FindClass find my class?](https://developer.android.com/training/articles/perf-jni#faq:-why-didnt-findclass-find-my-class)) is very limited, so some amount of trial-and-error may be required to get this right.

> The type of JBindingSession::_classLoaderObjects is a map without comparator. This leads to problems. Is it on purpose or is it a bug?
> (I added std::str_cmp for now, so that the key is the actual string and not the char const * pointer itself. In the future I would propose to use an enumeration for this.)

Yes, this is a bug! Good catch, will fix this.
Going forward I think it may be easier to just use std::string as keys in the map instead. I guess enums could be used too, but that will require further work, as some class names are defined with hardcoded strings (`"IOutStream"`, etc) while others have proper macro defines (`INSTREAM_CLASS`, etc).

> I managed to run JUnit tests on my Android device (that are failing for now). Is it a possibility to run JUnit tests (on a Simulator or on a Device) in a headless mode? Should you then get also the console output of the tests?

Yes, if you install Android Studio, you can run gradlew from the command line, including running unit tests.
Reference: [Test from the command line](https://developer.android.com/studio/test/command-line)

Reply:
> The meta code, that raises that exception is something like:
> ...
> Any ideas, why it happen?

Not sure, I'm afraid. Looks like there may be something fundamentally different between Java and Android ClassLoader implementations.

> Also, in you original code
> ...
> you don't convert `name` to dot-notation: java/lang/Object => java.lang.Object. Why is it working on Android anyway?

True, it looks like the Android ClassLoader happily accepts class names delimited with '/', even though the documentation states that the class name must be a binary name: [Binary names](https://developer.android.com/reference/java/lang/ClassLoader#binary-names). Not sure why.

In any case, I changed `findClass()` to use binary class names instead, and this worked just as well. So will use this solution instead.

Changes here:
- sevenzipjbinding: [ab40c13a483c0ca3d7504e04329726dcc14aa8f4](https://github.com/omicronapps/sevenzipjbinding/commit/ab40c13a483c0ca3d7504e04329726dcc14aa8f4)
- 7-Zip-JBinding-4Android: [9a2425513ff5c4a6278a267694f47a4409d9c93d](https://github.com/omicronapps/7-Zip-JBinding-4Android/commit/9a2425513ff5c4a6278a267694f47a4409d9c93d)